File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/tree/Problem1991.java
package com.keunwon.algorithm.tree;

import java.io.*;

public class Problem1991 {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    private static int n;
    private static Tree tree;

    public static void main(String[] args) throws IOException {
        input();
        tree.preorder(tree.root);
        bw.write(System.lineSeparator());

        tree.inorder(tree.root);
        bw.write(System.lineSeparator());

        tree.postorder(tree.root);
        bw.write(System.lineSeparator());

        bw.flush();
        bw.close();
        br.close();
    }

    private static void input() throws IOException {
        n = Integer.parseInt(br.readLine());
        tree = new Tree();

        for (int i = 0; i < n; i++) {
            char[] arr = br.readLine().replaceAll(" ", "").toCharArray();
            tree.add(arr[0], arr[1], arr[2]);
        }
    }

    private static class Node {
        char data;
        Node left;
        Node right;

        Node(char data) {
            this.data = data;
        }
    }

    private static class Tree {
        Node root;

        public void add(char data, char left, char right) {
            if (root == null) {
                root = new Node(data);

                if (left != '.') {
                    root.left = new Node(left);
                }
                if (right != '.') {
                    root.right = new Node(right);
                }
                return;
            }

            search(root, data, left, right);
        }

        public void search(Node root, char data, char left, char right) {
            if (root == null) {
                return;
            }

            if (root.data == data) {
                if (left != '.') {
                    root.left = new Node(left);
                }
                if (right != '.') {
                    root.right = new Node(right);
                }
                return;
            }

            search(root.left, data, left, right);
            search(root.right, data, left, right);
        }

        public void preorder(Node root) throws IOException {
            bw.write(root.data);

            if (root.left != null) { preorder(root.left); }
            if (root.right != null) { preorder(root.right); }
        }

        public void inorder(Node root) throws IOException {
            if (root.left != null) { inorder(root.left); }
            bw.write(root.data);
            if (root.right != null) { inorder(root.right); }
        }

        public void postorder(Node root) throws IOException {
            if (root.left != null) { postorder(root.left); }
            if (root.right != null) { postorder(root.right); }
            bw.write(root.data);
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem1991.kt
package com.keunwon.algorithm.baekjoon

/**
 * Title: 트리 순회
 * Level: 실버-1
 **/
class Problem1991 {
    private val answer = Array(3) { mutableListOf<Char>() }

    fun solution(arr: Array<Triple<Char, Char, Char>>): Array<String> {
        val root = arr[0].let { (target, left, right) ->
            Node(
                id = target,
                left = Node(left).takeUnless { left == '.' },
                right = Node(right).takeUnless { right == '.' },
            )
        }

        for (i in 1 until arr.size) {
            val (target, left, right) = arr[i]
            insert(root, target, left, right)
        }

        preOrder(root)
        inOrder(root)
        postOrder(root)
        return answer.map { it.joinToString("") }.toTypedArray()
    }

    private fun preOrder(node: Node?) {
        if (node == null) return

        answer[0].add(node.id)
        preOrder(node.left)
        preOrder(node.right)
    }

    private fun inOrder(node: Node?) {
        if (node == null) return

        inOrder(node.left)
        answer[1].add(node.id)
        inOrder(node.right)
    }

    private fun postOrder(node: Node?) {
        if (node == null) return

        postOrder(node.left)
        postOrder(node.right)
        answer[2].add(node.id)
    }

    private fun insert(node: Node, target: Char, left: Char, right: Char) {
        if (node.id == target) {
            node.left = Node(left).takeUnless { left == '.' }
            node.right = Node(right).takeUnless { right == '.' }
            return
        }
        node.left?.let { insert(it, target, left, right) }
        node.right?.let { insert(it, target, left, right) }
    }

    private data class Node(
        var id: Char,
        var left: Node? = null,
        var right: Node? = null,
    )
}

fun main() {
    val n = readLine()!!.toInt()
    val arr = Array(n) {
        readLine()!!
            .replace(" ", "")
            .toCharArray()
            .let { (a, b, c) -> Triple(a, b, c) }
    }

    Problem1991().solution(arr).forEach(::println)
}
